pubsub: start work on the channel api

reqshark opened the pull request:
#### fixes #31 and alters our pubsub API, ~~rendering it incompatible~~.

#### UPDATE (3/21/2015): API is compatible and extended:
*  the default sub socket will remain as it was: initialized with auto-subscription to *all-channels*.
*  `socket.chan()` calls internal check for the *all-channels* listener, removing it if necessary 

New methods:

### socket.chan(*Array*)

### socket.rmchan(*comma separated*)

the new API lets us do stuff like:
```js
var sub = nano.socket('sub')
sub.chan(['AAPL', 'MSFT', 'whatever'])

sub.on('message', function(msg){
  // only get msgs prefixed by those registered by .chan()
  console.log(String(msg))
})
``` 

~~Also I'd like to take advantage of `NN_SUB_UNSUBSCRIBE` like~~:
UPDATE (3/21/2015): We now take advantage of `NN_SUB_UNSUBSCRIBE`:
```js
sub.rmchan('MSFT', 'whatever')

sub.on('message', function(msg){

  // only get msgs prefixed 'AAPL' (following from snippet above)
  console.log(String(msg))
})
```

~~I haven't updated tests, nor added a handler for `chans` given to `opts` object at socket initialization~~:

```js
var nano = require('nanomsg')
var sub = nano.socket('sub', {
  chan: ['AAPL', 'MSFT', 'whatever']
})
```

~~I will follow up and commit those two things next if this API change meets with consensus/approval.~~

With that, let's open the forum to all opinions/arguments in favor or against this change.

nickdesaulniers commented on lib/index.js:
```diff
@@ -315,6 +316,26 @@ Socket.prototype.maxreconn  = opt('maxreconn');
+Socket.prototype.chan       = function(){
+  var args = arguments;
+  var channels = args.length;
+  while(--channels > -1) {
```
`while (--channels) {` is a common JS idiom, since `0` is false-y.

reqshark replied:
oh cool, but i think i need to include 0 here

nickdesaulniers replied:
Not if you subtract 1 from your array index, or use post decrement.
```
while (--i > -1) { arr[i]; }
```
compared to:
```
while (i--) { arr[i] }
```

reqshark replied:
sweet!

nickdesaulniers commented on lib/index.js:
```diff
@@ -315,6 +316,26 @@ Socket.prototype.maxreconn  = opt('maxreconn');
 
+/* chan and rmchan sockopt methods. only relevant for subscription sockets */
+Socket.prototype.chan       = function(){
+  var args = arguments;
```
Let's get rid of this additional variable.

reqshark replied:
Oh I did that since otherwise writing `arguments` makes this line quite long for readability, are you sure?
```js
if(nn.Setopt(this.binding, nn.NN_SUB, nn.NN_SUB_SUBSCRIBE, args[channels]) == 0)
```

nickdesaulniers commented on lib/index.js:
```diff
@@ -315,6 +316,26 @@ Socket.prototype.maxreconn  = opt('maxreconn');
+  var channels = args.length;
+  while(--channels > -1) {
+    if(nn.Setopt(this.binding, nn.NN_SUB, nn.NN_SUB_SUBSCRIBE, args[channels]) == 0)
+      this.channels[args[channels]] = 'active';
```
let's just assign this as a boolean instead of the string 'active'.  Also, please use `{}` with single statement if statements.

nickdesaulniers commented on lib/index.js:
```diff
@@ -315,6 +316,26 @@ Socket.prototype.maxreconn  = opt('maxreconn');
+  var args = arguments;
+  var channels = args.length;
+  while(--channels > -1) {
+    if(nn.Setopt(this.binding, nn.NN_SUB, nn.NN_SUB_UNSUBSCRIBE, args[channels]) == 0)
```
We should check that the channel is even set before calling into C++ unsubscribe.

reqshark replied:
good call

nickdesaulniers commented on src/node_nanomsg.cc:
```diff
@@ -37,11 +37,11 @@ NAN_METHOD(Socket) {
+        //if (nn_setsockopt(ret, NN_SUB, NN_SUB_SUBSCRIBE, "", 0) != 0) {
+            //return NanThrowError("Could not set subscribe option.");
+        //}
+    //}
```
Let's just remove this.

nickdesaulniers commented on src/node_nanomsg.cc:
```diff
@@ -136,7 +136,7 @@ NAN_METHOD(Setopt) {
   int option = args[2].integer;
 
-  if(option == NN_SOCKET_NAME){
+  if(option == NN_SOCKET_NAME || option == NN_SUB_SUBSCRIBE){
```
Hmm, so this isn't quite complete yet, right?

reqshark replied:
... good catch

reqshark commented:
here's a detailed summary of the last commit and what we're doing for sub sockets:

updated test files with subscription sockets:
* `test/inproc.js`
* `test/ipc.js`
* `test/sockoptapi.js`
* `test/tcp.js`
* `test/transform.js`

added test file demonstrating some of the new channel api:
* `test/chan.js`

removed test files:
* `test/getsockopt.js`
* `test/setsockopt.js`

these were duplicates since documented `sockopts` met test coverage plus some additional.

in `src/node_nanomsg.cc`, removed code:
*  `NAN_METHOD(Setsockopt)`
*  `NAN_METHOD(Getsockopt)`

and simplified existing methods by dropping support for `NN_SOCKET_NAME`:
* `NAN_METHOD(Setopt)`
* `NAN_METHOD(Getopt)`

added a new method to handle subscribe/unsubscribe:
* `NAN_METHOD(Chan)`

we're keeping socket option code restrained by separating the subscription handling concern out into its own small function. the rule of modularity asks us to do one thing and do it well.

Although socket channel registration is performed by `nn_setsockopt()`, @nickdesaulniers and I discussed and found shifting the operation into `NAN_METHOD(Chan)` reduces C pointer type complexity ordinarily arising from an encapsulated `Setsockopt()` method. This alternative also provides a concise and complete interface underlying API calls to `socket.chan()` and `socket.rmchan()`.

with fewer tests there are exactly 800 nyan-cat, all green :palm_tree:

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,38 @@ function Socket (type, opts) {
+  if (opts.hasOwnProperty('chan')) {
+    if (typeof opts.chan == 'string') {
+      opts.chan.split(',').forEach(subscriber.bind(this))
+    } else {
```
This `else` branch should be an `else if (Array.isArray(opts.chan)) {`

reqshark replied:
OK but if that thing is neither String nor Array then I want that forEach to throw the process

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,38 @@ function Socket (type, opts) {
+      opts.chan.split(',').forEach(subscriber.bind(this))
+    } else {
+      opts.chan.forEach(subscriber.bind(this))
     }
```
In both cases of the conditional, the code would be more robust if we first called `.trim()` on what was being passed to subscriber.  In the case of splitting, we should filter out empty elements, after trimming.  If you map trim across all elements, then filter on the first branch, that would be nice.

reqshark replied:
oh nice! started to think about that too after I saw how that string looked

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,38 @@ function Socket (type, opts) {
+
+  /* subscription handling at initialization */
+  if (opts.hasOwnProperty('chan')) {
+    if (typeof opts.chan == 'string') {
```
Let's use strict equality everywhere, always.

reqshark replied:
word

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,38 @@ function Socket (type, opts) {
+    if (nn.Chan(this.binding, nn.NN_SUB_SUBSCRIBE, chan) != -1) {
+      this.channels[chan] = true;
+    } else {
+      console.log('sub err')
```
Let's log this to stderr with console.error.

reqshark replied:
I was thinking maybe we should throw here actually.

One thing that will come up in another PR (forthcoming related to `nn_shutdown()`), is that we'll need to decide about dynamic re-subscription, according to [folks on the mailing list](http://www.freelists.org/post/nanomsg/Question-on-pubsub-device,5) it may not be possible, and backing that idea up, I'm seeing some unverified issues with latent calls after `socket.rmchan()` on a specific topic.

nickdesaulniers commented on lib/index.js:
```diff
@@ -372,6 +394,28 @@ function opt (option) {
 };
 
+/* chan and rmchan sockopt methods. only relevant for subscription sockets */
+Socket.prototype.chan       = function(){
```
One of the issues I have with the current API is that during socket construction, you can pass either a comma separated string, or an array.  After construction, you can call the chan method with only an array.

Since you might want to filter on chans that contain a comma, splitting on the comma in the constructor seems kind of brittle.  Rather than know about that edge case, why don't we eliminate it and require library consumers to have to always pass an array of strings?

reqshark replied:
> After construction, you can call the chan method with only an array.

well `arguments` is certainly array-like. after socket construction `chan()` accepts variable arity, so it's just comma separated parameters

nickdesaulniers replied:
Then you can spread them out:

```js
chan.apply(this, Array.prototype.slice.call(arguments));
```

nickdesaulniers replied:
Further, if we only take an array of strings, then there's no ambiguity about whitespace, and we don't need to filter or map the input.

reqshark replied:
ok i was holding back since `nano.socket('sub', { chan: [''] })` is unacceptable... 
... as far as style goes.

but since you came up with that idea of `''` as the defualt where we run a `socket.chanrm('')` when passing in topic lists, let's definitely do that

nickdesaulniers commented on lib/index.js:
```diff
@@ -372,6 +394,28 @@ function opt (option) {
+Socket.prototype.chan       = function(){
+  var args = arguments;
+  var channels = args.length;
+  while(channels--) {
```
Would prefer that variables used for the sole purpose of iteration be named `i`, `j`, or `k`.

reqshark replied:
I personally would understand the code, and it would look cleaner to me with less of it, but what about someone new to the project? 

Is short `var` more legible than a descriptive `var`?

As I recall ZeroMQ is somewhat opinionated about that, of course that's not necessarily the style of libnanomsg, whose naming conventions are more aligned with POSIX standards than probably anything else.

nickdesaulniers commented on lib/index.js:
```diff
@@ -372,6 +394,28 @@ function opt (option) {
+  while(channels--) {
+    if (this.channels[args[channels]] &&
+      nn.Chan(this.binding, nn.NN_SUB_UNSUBSCRIBE, args[channels]) != -1) {
+        this.channels[args[channels]] = false;
```
Maybe we should be deleting the key from the obj, rather than setting it false?  Eh, I'm on the fence.  Thoughts?

reqshark replied:
definitely, good idea, I want to use the `delete` operator

nickdesaulniers commented on test/chan.js:
```diff
@@ -0,0 +1,75 @@
+      case 0: return t.equal(msg, msgs['m' + msgqty]); //msgs.m0, 'hello world'
+      case 1: return t.equal(msg, msgs['m' + msgqty]); //msgs.m1, 'hello world'
+      case 2: return t.equal(msg, msgs['m' + msgqty]); //msgs.m2, 'hello world'
+      case 3: return t.equal(msg, msgs['m' + msgqty]); //msgs.m3, 'bar world'
```
just make msgs an array, then here you can `return t.equal(msg, msgs[msgqty])`

nickdesaulniers replied:
ah nevermind, I see that's what removeHello is doing

nickdesaulniers replied:
bump

nickdesaulniers commented on test/chan.js:
```diff
@@ -0,0 +1,75 @@
+    m2:'hello world',
+    m3:'bar world',
+    m4:'foo world'
+  }
```
we should add a message that is not a chan we're subscribed too and explicitly fail if that message is encountered.

reqshark replied:
for the test ya i agree, that was my purpose for writing the `removeHello` function

nickdesaulniers commented on test/inproc.js:
```diff
@@ -10,7 +10,7 @@ test('inproc socket pub sub', function (t) {
 
     var pub = nano.socket('pub');
-    var sub = nano.socket('sub');
+    var sub = nano.socket('sub', {chan:''});
```
While it's nice to see you're being explicit about not subscribing to any chans, this makes it look like the second arg to nano.socket is required.  Let's remove the explicitly empty options argument, that way users reading the tests for usage don't think that they are required to construct a meaningless object.

reqshark replied:
from the beginning with these experiments [my philosophy was to set options early](https://github.com/reqshark/nanomsg.iojs/tree/a49cacc15bb8c6ca195b50d2ea780f4f92bec3fd#philosophy-setsockopts-early).

i think it's better practice than calling `.chan()` later, but sure, we can do that

nickdesaulniers commented on test/mksymbols.rb:
```diff
@@ -42,4 +42,3 @@
 #{o}
 ];
 FOO
-
```
Let's try to keep new lines at the end of files, see: http://stackoverflow.com/questions/729692/why-should-files-end-with-a-newline

reqshark replied:
weird i didn't meant to do that.

nickdesaulniers replied:
bump

reqshark replied:
github atom editor doesn't let me fix this file. that's what i've been using lately.

i think there is a line when you pull it down. whatever i use vim to fix that

nickdesaulniers replied:
Use the git command line and remove it from the patch.  http://stackoverflow.com/questions/5020305/how-can-i-remove-a-change-in-1-file-in-1-commit

nickdesaulniers commented:
Looks good!  A few minor nits, the biggest being that I think we should always expect an array of strings.

This will require a major version bump, since we're removing the getsockopt and setsockopt api, and requiring you to go through the getters/setters.

Patches that remove more code than they add :+1: :+1: :+1: Thanks @reqshark !

nickdesaulniers commented on src/node_nanomsg.cc:
```diff
@@ -57,86 +51,14 @@ NAN_METHOD(Close) {
-    ret(NanNew<Number>(
-        nn_setsockopt(S, level, option, &optval, sizeof(optval))));
-  }
+  ret(NanNew<Number>(nn_setsockopt(S, level, option, &optval, sizeof(optval))));
```
make sure you run `clang-format -style=Mozilla -i <file>` on this!

reqshark replied:
ok will do before i update my changes. 

btw, i'll try to preserve the commit history, the reason it's lost before this point is that merge conflict yesterday..

reqshark replied:
`clang-format -style=Mozilla -i src/node_nanomsg.cc` approved!   :rocket:

nickdesaulniers commented:
Please edit your commit messages to mention that they `Fixes #67, Fixes #31`.  Then, GH will auto close those issues when the patch is merged.

reqshark commented:
> Please edit your commit messages to mention that they Fixes #67, Fixes #31. Then, GH will auto close those issues when the patch is merged.

yep before merging, I just figured if i needed to rebase or whatever i didn't want to keep re-referencing phantom commits

nickdesaulniers commented on test/inproc.js:
```diff
@@ -15,6 +15,8 @@ test('inproc socket pub sub', function (t) {
     var addr = 'inproc://pubsub';
     var msg = 'hello world';
 
+    sub.chan('');
```
Ah, I meant can we remove all references to the `''` chan from tests that don't test filtering on a sub socket?  Will the sub socket fail to work, since I think we always have to at least subscribe to the empty chan?  If so, since we can have multiple subscription, then we should call `this.chan('')` during socket construction if it's a sub socket.  Otherwise, it's silly to force users to have to explicitly subscribe to the empty channel just to use sub sockets.

reqshark replied:
> can we remove all references to the '' chan from tests that don't test filtering on a sub socket?

that would be useful. maybe if we run a setTimeout or Interval after sub socket is created that checks for the existence of any length on `this.channel`. 

If we did something like that it would make more sense to only accept arrays for topics, since we could just check length on that property.

well if you have a better implementation idea in mind I'm all ears, but the main thing is that once we've subscribed to `''`, there's no point in registering subsequent subscription topics, since these plus all others would get filtered into the `message` event callback

nickdesaulniers replied:
> that would be useful. maybe if we run a setTimeout or Interval after sub socket is created that checks for the existence of any length on this.channel.

Then you run into the polling problem.  What duration do you use? If too long then you miss messages, if too short then you are subject to v8/machine perf.  Too brittle.

So, if you don't subscribe to any channels explicitly, sub sockets don't work (can we confirm)?

> once we've subscribed to '', there's no point in registering subsequent subscription topics

Can we make `''` the default, then the first time someone calls `chan`, can we `rmchan('')`?

reqshark replied:
> Can we make '' the default, then the first time someone calls chan, can we rmchan('')?

excellent idea.

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +106,40 @@ function Socket (type, opts) {
+        return i.trim();
+      }).forEach(subscriber.bind(this))
+    } else if (Array.isArray(opts.chan)) {
+      opts.chan.forEach(subscriber.bind(this))
```
in both cases, we should map trim, then filter empty elements.  This prevents `chan: ['' ' a', 'b ', ' c ']` and `chan: ", a,b , c "`.

nickdesaulniers commented on lib/index.js:
```diff
@@ -4,6 +4,16 @@ var util = require('util');
+  param           : 'parameter type err: ',
+  chan            : 'setting channel subscription string: ',
+  rmchan          : 'removing channel subscription string: ',
+}
```
There's really no need for this, just put the strings in place.  DRY has its place, but not with error message substrings.

reqshark replied:
ya the need is all these additional calls to `new Error('...` are longer than 81 chars per line.

I'd rather one-line new Errors, but if you insist.

well, hey I had some other repeated errs in mind too, it's not just those.

eh, I can take it out I guess, but you'd really rather see calls to `new Error` broken into multiple lines, or worse: trailing off beyond 80 chars to the right of the screen?

nickdesaulniers replied:
We'll let the formatter figure it out, when it lands.  I don't have my heart set on 80 char line limits.

reqshark replied:
> I don't have my heart set on 80 char line limits.

well I read code on the train from my phone..

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +106,40 @@ function Socket (type, opts) {
+  }
+
+  function subscriber(chan) {
+    if (nn.Chan(this.binding, nn.NN_SUB_SUBSCRIBE, chan) !== -1)
```
curly braces

reqshark commented:
I have mixed feelings about how we're handling the default string registration and then special `rmchan()` stuff when that's the case. it's taking me like 30 lines of code starting from around 100 and it seems overly complicated to me... but that's the simpliest way i saw solving it... 

actually maybe it's not that bad, i just took a second look after stepping away from the code...

@nickdesaulniers please review

nickdesaulniers commented on test/tcp.js:
```diff
@@ -15,6 +15,8 @@ test('tcp socket pub sub', function (t) {
     var sub = nano.socket('sub');
 
+    sub.chan('');
+
```
can be removed.

nickdesaulniers commented on lib/index.js:
```diff
@@ -372,6 +404,32 @@ function opt (option) {
 };
 
+/* chan and rmchan sockopt methods. only relevant for subscription sockets */
+Socket.prototype.chan       = function(list){
```
list here is an unused variable

reqshark replied:
oof, i think i pushed the wrong state. haha it's on my home computer. anyway, `this.register` function iterates the list. i'll update that to reflect list as an array as discussed earlier

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,48 @@ function Socket (type, opts) {
+      throw new TypeError('chan requires an Array');
     }
+  } else {
+    if (type === 'sub') {
```
this if can merge with the line above it `} else if (...) {`

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,48 @@ function Socket (type, opts) {
+  } else {
+    if (type === 'sub') {
+      this.register(''); //default topic is an empty string
+      setImmediate(function(){ this.chanDefault = true; }.bind(this));
```
What is this?

reqshark replied:
the `chanDefault` flag. 

waiting a turn of the event loop before setting `this.chanDefault` to true is a guarantee that `register()`'s check on line 122, occurs before `this.chanDefault` gets set to true.

it's excessive and we don't need `this.chanDefault`, if we set the empty string on `this.channels`, like we discussed on IRC this morning

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,48 @@ function Socket (type, opts) {
+      return this.channels[chan] = true;
+    }
+    return this.emit('error', new Error( nn.Err() + ' : ' + chan));
+  }
```
this functionality is definitely duplicated by `this.chan`

reqshark replied:
ya i pushed the wrong version of the code, it's supposed to call `register()`

nickdesaulniers commented on lib/index.js:
```diff
@@ -372,6 +404,32 @@ function opt (option) {
+    if (nn.Chan(this.binding, nn.NN_SUB_SUBSCRIBE, arguments[i]) !== -1) {
+      this.channels[arguments[i]] = true;
+    } else {
+      this.emit('error', new Error( er.op + er.chan + arguments[i] + '\n'));
```
this error handler with throw an error, er.op and er.chan

reqshark replied:
not the code i meant to push.

nickdesaulniers commented on lib/index.js:
```diff
@@ -372,6 +404,32 @@ function opt (option) {
+      if (nn.Chan(this.binding, nn.NN_SUB_UNSUBSCRIBE, arguments[i]) !== -1) {
+        delete this.channels[arguments[i]];
+      } else {
+        this.emit('error', new Error( er.op + er.rmchan + arguments[i] + '\n'));
```
this error handler will throw an error.  Removed the definitions but did not replace the usages.

reqshark replied:
dito

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,48 @@ function Socket (type, opts) {
+  /*  subscription filter control */
+  this.channels     = {};
+  this.register     = register;
+  this.chanDefault  = false;
```
let's remove this, and use the ninja variable technique we discussed on irc.

nickdesaulniers commented on test/chan.js:
```diff
@@ -0,0 +1,69 @@
+
+test('adding and removing subscription channels', function (t) {
+
+  //t.plan(7);
```
don't we need to list the number of plans?  if not, this can be removed

reqshark replied:
ok. yes, we either need to list plans or call `t.end()`

nickdesaulniers commented on test/chan.js:
```diff
@@ -0,0 +1,69 @@
+    pub.close();
+    t.end()
+  }
+});
```
we should add a test to handle the default subscriptions, since we're doing extra logic there. more thoughts on that in a few mins.

nickdesaulniers replied:
The test should:
* show that constructing a socket without explicitly subscribing to any channel receives at least 2 messages of different prefixes.
* show that constructing a socket without explicitly subscribing to any channel during construction, but subscribes after construction using chan receives correctly prefixed messages but not others.
* show that constructing a socket explicitly subscribing to a channel during construction receives the appropriately prefixed messages but not others.
* show that constructing a socket explicitly subscribing to two or more channels during construction, later rmchan all but one, will stop receiving messages to channels no longer subscribed to.
* what happens when we rm all chans?

reqshark replied:
ok but

> * show that constructing a socket without explicitly subscribing to any channel receives at least 2 messages of different prefixes.

seems like redundant coverage in `test/ipc.js`, `test/inproc.js`, `test/tcp.js`

nickdesaulniers commented on test/chan.js:
```diff
@@ -0,0 +1,69 @@
+    var msg = String(buf);
+
+    //increment msgqty when a msg is received (happening about every 100ms)
+    //we'll call finish() on the 5th msg received (case 4, see line 68)
```
I wouldn't recommend putting explicit line numbers, it's brittle and prone to breakage as the code changes.

reqshark commented:
sorry still WIP right now, I'll update with another note when it's ready for the final review..

nickdesaulniers commented on lib/index.js:
```diff
@@ -96,16 +96,42 @@ function Socket (type, opts) {
+      return this.channels[chan] = true;
     }
+    return this.emit('error', new Error( nn.Err() + ' : ' + chan));
+  }
```
Can we make register a method of the prototype?
```js
Socket.prototype._register(chan) {};
```
Let's make each statement it's own line.
Let's remove return values since they're not used by any callers.
```js
if (this.chan...) {
} else if (nn.Chan) {
} else {}
// no explicit return statements
```

After this is fixed, the PR looks good to merge.  Now that Travis is hooked up, I trust it enough that I don't have to pull down the patch and test it locally! :rocket: :guitar:

nickdesaulniers commented on lib/index.js:
```diff
@@ -372,6 +398,27 @@ function opt (option) {
+      return this.emit('error', new Error( nn.Err() + ' : ' + chan));
+    }
+  };
+}
```
Actually, the return values here are also pretty useless, let's remove explicit return statements and just have the innermost if be an if-else.

nickdesaulniers commented:
> fixes #31 and alters our pubsub API, rendering it incompatible.

While it doesn't change the pubsub API, it does remove the getsockopt and setsockopt api, and thus will require a major version bump.

nickdesaulniers commented:
Let's focus on getting this wrapped up.  There's just one last line note with a few nits, then this is good to land! :rocket: :guitar:

reqshark commented:
ok let me just fix this merge conflict and add those last nits, I'll sort this out shortly
